flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala

```diff
@@ -177,9 +181,92 @@ object TableSourceUtil {
     }
     if (selectedFields.isDefined) {
       // filter field names and types by selected fields
-      (selectedFields.get.map(fieldNames(_)), selectedFields.get.map(fieldTypes(_)))
+      fieldNames = selectedFields.get.map(fieldNames(_))
+      fieldTypes = selectedFields.get.map(fieldTypes(_))
+    }
+    typeFactory.buildRelNodeRowType(fieldNames, fieldTypes)
+  }
+
+  /**
+    * Returns schema of the selected fields of the given [[TableSource]].
+    *
+    * @param typeFactory Type factory to create the type
+    * @param fieldNameArray Field names to build the row type
+    * @param fieldDataTypeArray Field data types to build the row type
+    * @param watermarkSpec Watermark specifications defined through DDL
+    * @param selectedFields The indices of all selected fields. None, if all fields are selected
+    * @param streaming Flag to determine whether the schema of a stream or batch table is created
+    * @return The row type for the selected fields of the given [[TableSource]], this type would
+    *         also be patched with time attributes defined in the give [[WatermarkSpec]]
+    */
+  def getSourceRowType(
```

Comment:
Could you move this out of `TableSourceUtil`? The name implies that this util should work with some `TableSource`

Reply:
Thanks, i can't move it out because another method `def getSourceRowType` reference it in `TableSourceUtil`.